src/combinators/mod.rs

```diff
@@ -0,0 +1,135 @@
+use crate::*;
+use std::time::Duration;
+use std::{future::Future, time::Instant};
+use tokio::{self, select, time};
+
+pub fn map<E, I, O, F, Fut>(source: E, mut f: F) -> Eventual<O>
+where
+    E: IntoReader<Output = I>,
+    F: 'static + Send + FnMut(I) -> Fut,
+    I: Value,
+    O: Value,
+    Fut: Send + Future<Output = O>,
+{
+    let mut source = source.into_reader();
+
+    Eventual::spawn(|mut writer| async move {
+        while let Ok(v) = source.next().await {
+            writer.write(f(v).await);
+        }
+    })
+}
+
+pub fn timer(interval: Duration) -> Eventual<Instant> {
+    Eventual::spawn(move |mut writer| async move {
+        loop {
+            writer.write(Instant::now());
+            time::sleep(interval).await;
+        }
+    })
+}
+
+// TODO: Put this in a macro to de-duplicate A/B and support more arguments.
+pub fn join<A, Ar, B, Br>(a: Ar, b: Br) -> Eventual<(A, B)>
```

Comment:
Yes, that's the plan. I typically implement things by hand once before doing a macro because it makes it easier to separate correctness from generating correct syntax in my head. Good to think about one thing at a time.